### PR TITLE
i244: now by default loads sample/ data files before secret

### LINKED ISSUE #244: Provide a way to add sample data files to the test data files
**Is your feature request related to a problem?** 

When shadowing there is a step required to create test cases data files.
The sample files should be tested before the judge's data files (in secret/)

Here is the workaround:

Sample and Secret data files need to be renamed and 
after renaming the sample files are copied to the secret directory.

Merge the “sample data” into the “secret data” folder and load ALL data
files into PC2 (Kattis runs against both the samples and the secret data).
There is a Python3 script in the NAC GitHub repo that can be used to do
this;  it takes a single argument -- the name of the “cdp” folder which
contains the config folder which contains the problem package (problem
folders and a single “problem.yaml” file).  Warning:  the script MOVES
the sample data into the secret data folder, prepending “sample__”
to all sample data files and “secret__” to all secret data files.
This produces a single folder, “secret”, containing all the data files in
proper alphabetical order -- but it destroys the original cdp structure.

Explanatory note:  data files are required to be run in alphabetical
order, and Kattis runs all sample files before any secret files

Source: NADC 2021 PC2 Shadow Configuration Verification Checklist

**Feature Description**: 

Using --load (from yaml) automatically load sample files before loading secret files for each problem.

For work arround:

Copy sample files to data/secret directory, rename all data files so that the sample files will
be used before the secret files.

Copy each .in and .ans file from sample data directory  (config\<shortdirname>\data\sample) to secret directory
(config\<shortdirname>\data\secret) with a new name, the new file name should be the sample file name
prepended with sample__ .   Ex. data/sample/sumit.in copy to data/secret/sample__sumit.in

Rename all secret files, prepend secret_ to each secret__, ex data/secret/sumit4.ans to data/secret/secret__sumit4.ans

By default renames and copy files.
Provide two options: 
 - output a script to change files
 - list files to be changed and show counts of files to be created/renamed.

**Have you considered other ways to accomplish the same thing?** 

There are scripts to rename and copy samples into the secret directory

**Do you have any specific suggestions for how your feature would be ***implemented*** in PC^2?** 

Yes, create a new tool that will perform the renaming and copying.

**Additional context**:

### PR DESCRIPTION
**Pull request by lane55**
### Description of what the PR does

On --load (contest yaml load) will automatically load sample data files before secret data files.

### Issue which the PR fixes

#244 

### Environment in which the PR was developed (OS,IDE, Java version, etc.)

Microsoft Windows [Version 10.0.19044.2130]

java version "1.8.0_121"
Java(TM) SE Runtime Environment (build 1.8.0_121-b13)
Java HotSpot(TM) 64-Bit Server VM (build 25.121-b13, mixed mode)

### Precise steps for _testing_ the PR (i.e., how to demonstrate that it works correctly)

Before: Load sample mini contest, for the Sumit (only) problem there are 7 test cases

After: Load sample mini contest, for the Sumit (only) problem there are 10 test cases
and includes the 3 samples

### CODE REVIEW

File: src/edu/csus/ecs/pc2/imports/ccs/ContestSnakeYAMLLoader.java
Comment (johnbrvc) [inline src/edu/csus/ecs/pc2/imports/ccs/ContestSnakeYAMLLoader.java:2822]: Shouldn't a check be made to make sure that inputFileNames.length == answerFileNames.length ?

File: src/edu/csus/ecs/pc2/imports/ccs/ContestSnakeYAMLLoader.java
Comment (johnbrvc) [inline src/edu/csus/ecs/pc2/imports/ccs/ContestSnakeYAMLLoader.java:2755]: Since you wrote a routine to load datafiles from an arbitrary directory, why not use loadDataFiles(problem, dataFiles, answerFiles, dataFileBaseDirectory, loadExternalFile); and fix up loadDataFiles so it can work for both?  I understand there will have to be some small changes to the stuff that sets up the dataFiles and answerFiles arraylists, but it would make it slightly cleaner in the long run.

File: src/edu/csus/ecs/pc2/imports/ccs/ContestSnakeYAMLLoader.java
Comment (johnbrvc) [inline src/edu/csus/ecs/pc2/imports/ccs/ContestSnakeYAMLLoader.java:2833]: Should a check be made here for idx == 0 to set the problem.setDataFIleName(xxx), problem.setAnswerFileName(xxx) like it is for the secret data, and then, don't do it for the secret data?  Basically, the test should probably be based on the number of test cases already added to the problem (problem.getNumberTestCases()) as opposed to the index of the test case being added.  This way, it will work for both sample and secret data and only add the first one;  the setDataFileName() sets the single name that appears in the problem display as I recall, and if sample data is in the list, THAT should be the file shown for consistency.

**Review by johnbrvc — COMMENTED** [review]
This is an extremely useful and needed fix, and we should try to get it approved and merged ASAP (maybe for ACPC, assuming they are even using the test-case mechanism that PC2 has).

File: samps/contests/mini/config/contest.yaml
Comment (clevengr) [inline samps/contests/mini/config/contest.yaml:41]: I do not understand this change.  The text comment (3 lines up) says "default is true", but BOTH of the lines just above are commented-out.  Was the intent that the second line should have been uncommented (making the comment correct)?  Otherwise, I'm totally mystified as to why there are two lines (39 and 40), both comments, giving opposite values.

File: src/edu/csus/ecs/pc2/imports/ccs/ContestSnakeYAMLLoader.java
Comment (clevengr) [inline src/edu/csus/ecs/pc2/imports/ccs/ContestSnakeYAMLLoader.java:270]: Totally minor, picky point:  java method names should start with a lower-case letter.

File: samps/contests/mini/config/contest.yaml
Comment (clevengr) [reply samps/contests/mini/config/contest.yaml:41]: Oh, unless the comment means that in the CODE the "default is true", and this sample YAML provides two options, both of which are currently commented-out but where the user can choose to uncomment one of them to override the code default?   That would make sense... if the comment explained it.  If that's the intent, I think the comment should be expanded to clarify the intent...

File: src/edu/csus/ecs/pc2/imports/ccs/ContestSnakeYAMLLoader.java
Comment (clevengr) [inline src/edu/csus/ecs/pc2/imports/ccs/ContestSnakeYAMLLoader.java:2735]: These lines of code are confusing; they get the boolean value `loadSamples` and the name of the `sampleDataDirectory`, but then those values are not used in the following lines (which use `dataFileBaseDirectory` instead, and make no use of `loadSamples`).   It would be much clearer if these lines were moved down just in front of line 2756, which is the place where these values are actually used.

File: src/edu/csus/ecs/pc2/imports/ccs/ContestSnakeYAMLLoader.java
Comment (clevengr) [inline src/edu/csus/ecs/pc2/imports/ccs/ContestSnakeYAMLLoader.java:2827]: Hmm... this line of code references element `[idx]` in both the `inputFileNames` and `answerFilesNames` arrays.  The first one (`inputFileNames`) has been sorted (line 2823, above), but the second one has not.  Doesn't that mean that the elements at `[idx]` in the two arrays are no longer guaranteed to be corresponding?

File: src/edu/csus/ecs/pc2/imports/ccs/ContestSnakeYAMLLoader.java
Comment (clevengr) [inline src/edu/csus/ecs/pc2/imports/ccs/ContestSnakeYAMLLoader.java:2821]: Since this method is being used to load ONLY sample data files, for clarity I would suggest a better method name would be `loadSampleDataFiles`.  Although actually I agree with JohnB's earlier comment that the most logical thing to do would be to use the same method for loading both samples and secrets.

File: src/edu/csus/ecs/pc2/imports/ccs/ContestSnakeYAMLLoader.java
Comment (clevengr) [reply src/edu/csus/ecs/pc2/imports/ccs/ContestSnakeYAMLLoader.java]: Although I see that this is also the way it was done when only the "secret" files were being loaded, so (presumably?) it works... although I'm not sure how.

**Review by clevengr — COMMENTED** [review]
I'd like to see responses to the questions/suggestions that both JohnB and I have made.

File: src/edu/csus/ecs/pc2/imports/ccs/ContestSnakeYAMLLoader.java
Comment (johnbrvc) [reply src/edu/csus/ecs/pc2/imports/ccs/ContestSnakeYAMLLoader.java]: Yes, I followed the logic through, and unless I missed something, it would only work if the directory is already in sorted order.  There is also code in that "idx" loop that creates 2 arrays (dataFiles and answerFiles) for use just below in computing "problemDataFiles", which is passed into loadCCSProblemFiles.  It turns out that problemDataFiles is never used by the caller.  Evidently it WAS used at one point for the ClicsOutputValidator, but it is no longer used in there, so that code is dunsel.  Note, however, that the code that calculates dataFiles and answerFiles does go out of its way to compute the CORRECT answer file name based on the input file name. `String answerShortFileName = inputFileNames[idx].replaceAll(".in$", ".ans");`  However, as I said, it is for naught, since those arrays serve no useful purpose.  Similar code should probably be used to form the answer file name before the call to `problem.addTestCaseFilenames()`.  Perhaps:

`problem.addTestCaseFilenames(inputFileNames[idx], inputFileNames[idx].replaceAll(".in", ".ans"));`

File: src/edu/csus/ecs/pc2/imports/ccs/ContestSnakeYAMLLoader.java
Comment (johnbrvc) [inline src/edu/csus/ecs/pc2/imports/ccs/ContestSnakeYAMLLoader.java:2760]: Is there any particular reason we can't do the refactor now since it should be pretty simple to do?

**Comment by johnbrvc** [comment]
There are still a couple of outstanding questions that need answering regarding this PR.
See above comments.

**Comment by johnbrvc** [comment]
What about the  idx == 0 and the setDataFileName() and setAnswerFileName() comment I made above.  This is a bug created by the addition of the sample files being read.  The wrong value will be set if loading both sample and secret data.  The secret data first test case names will overwrite the sample data first test case names.

**Comment by lane55** [comment]
> What about the idx == 0 and the setDataFileName() and setAnswerFileName() comment I made above. This is a bug created by the addition of the sample files being read. The wrong value will be set if loading both sample and secret data. The secret data first test case names will overwrite the sample data first test case names.

> What about the idx == 0 and the setDataFileName() and setAnswerFileName() comment I made above. This is a bug created by the addition of the sample files being read. The wrong value will be set if loading both sample and secret data. The secret data first test case names will overwrite the sample data first test case names.

This is very old code that works fine, there is no bug.

This code are not related to filenames that Executeable uses unless
the input type is File I/O.  You can look in Executable to see where
thses fields are used.

Since we do not use File I/O input type, these fields are not used.

File: src/edu/csus/ecs/pc2/imports/ccs/ContestSnakeYAMLLoader.java
Comment (clevengr) [reply src/edu/csus/ecs/pc2/imports/ccs/ContestSnakeYAMLLoader.java]: Not something that should hold up approval; just sayin'....

File: src/edu/csus/ecs/pc2/imports/ccs/ContestSnakeYAMLLoader.java
Comment (clevengr) [reply src/edu/csus/ecs/pc2/imports/ccs/ContestSnakeYAMLLoader.java:2735]: Again, not something that should hold up approval... I just don't care for adding code that's confusing (even if it executes correctly) when it could be avoided...

**Comment by johnbrvc** [comment]
> > What about the idx == 0 and the setDataFileName() and setAnswerFileName() comment I made above. This is a bug created by the addition of the sample files being read. The wrong value will be set if loading both sample and secret data. The secret data first test case names will overwrite the sample data first test case names.
> 
> > What about the idx == 0 and the setDataFileName() and setAnswerFileName() comment I made above. This is a bug created by the addition of the sample files being read. The wrong value will be set if loading both sample and secret data. The secret data first test case names will overwrite the sample data first test case names.
> 
> This is very old code that works fine, there is no bug.
> 
> This code are not related to filenames that Executeable uses unless the input type is File I/O. You can look in Executable to see where thses fields are used.
> 
> Since we do not use File I/O input type, these fields are not used.

Well, I still believe it is a bug.  The **_Edit Problem_** dialog will not show the FIRST file in the judging data, rather, it will show the first file in the "secret" judging data, instead of the first file in the sample judging data.

load

In the picture, the files in the **_Input Data File_** and **_Answer File_** fields specify the first case of secret data,
not sample data, which would have been "1.in" and "1.ans"  (see WF Dhaka "mosaic" problem).

This can lead to confusion when looking at the **_Edit Problem_** dialog to see the name of the "first" file.  Even though it is not used per se', it is a visual aid to the user as to what judge data is being used.  Basically, it now works (looks) differently than it used to, since the files are being read from two different places.

While this may not be cause for stalling the PR, another issue should probably be created as a "someday" to make those fields reflect reality.  Perhaps in the "refactor" of sharing the loadDataFiles() routine?

File: src/edu/csus/ecs/pc2/imports/ccs/ContestSnakeYAMLLoader.java
Comment (clevengr) [reply src/edu/csus/ecs/pc2/imports/ccs/ContestSnakeYAMLLoader.java]: @johnbrvc  The original issue which I pointed out -- the fact that `inputFileNames` was being sorted while `answerFileNames` was not -- has been fixed (by the addition of a call to `Arrays.sort(answerFileNames)`.  However, I'm not sure whether to mark this conversation "resolved" because it appears maybe you introduced an addition issue in your comment.  What's your take -- does the addition of the call to `Arrays.sort()` fix the problem, or is there still another problem (besides "dunsel code") ?

**Comment by clevengr** [comment]
@lane55  said:
> This is very old code that works fine, there is no bug.

It may be "very old code", but it (no longer) "works fine":  as @johnbrvc points out, the addition of the new code which adds the sample data files causes the Edit Problem GUI to display incorrect information.  This in my mind is clearly a bug that's been introduced with the addition of including the sample data files and should be fixed (in other words, the code that adds sample data files is not doing everything correctly that it should).

**Comment by clevengr** [comment]
@lane55 said:

> This code are not related to filenames that Executeable uses unless the input type is File I/O. You can look in Executable to see where thses fields are used.
> 
> Since we do not use File I/O input type, these fields are not used.

It's completely wrong to say "we do not use File I/O".  First of all, most if not all the Sample Contests we supply are configured to use File I/O by default (something that bugs me every time I load SumitHello, for example).  Secondly, even if we didn't provide any Sample Contests (or if we changed them to default to StdIn I/O),  we still provide the option to users to select File I/O.  As long as we provide that option, we should not be including code updates that will display incorrect data if a user selects that mode.  

IMHO, this issue needs to be fixed before this PR can be approved.  (I know you don't want to hear that, but c'mon...  including code that displays wrong data if a user selects a PC2-provided mode?)

**Comment by johnbrvc** [comment]
I think we have to revisit this PR since its functionality is VERY important and automates the loading of a new contest.

**Comment by lane55** [comment]
That Edit Problem existing "bug" should not stop this from moving forward.   If it si critical then open an issue.

The Finals mosaic problem would have been judged correctly, because the fields in the example in Edit Problem
are **not used**.  Those names are not used in a Stdin data source.

**Comment by johnbrvc** [comment]
> 

IMHO since this is such a simple fix, is it really worth going through the whole issue/pr/merge process to address it?  Why not just add it as a CI now with its own commit.  With this PR, it would work "differently" than it used to - and that's my issue with it.  Previously it used to show the first "secret" test-case file name.  Now it will show the (N+1)th test case, where N is the # of sample test cases.  Again, IMHO.

**Comment by lane55** [comment]
I see a lot of testing work to fix this bug, that is why I suggested a new Issue - to define the fix and provide testing for this fix.

**Review by johnbrvc — APPROVED** [review]
Since the functionality in this PR is very much needed, I don't think the points raised in the comments should block the approval.

**Comment by clevengr** [comment]
Hmm... the PR has two approvals, but it can't be merged because of conflicts in the ContestSnakeYAMLLoaderTest file.  @lane55 : can you please look at what those conflicts are and resolve them?

**Review by clevengr — APPROVED** [review]
I could swear I indicated "Approve" on my last review...

**Review by clevengr — APPROVED** [review]

**Pull request merged**
